File: app/node_client.py
"""中心服务作为「测速客户端」去访问节点 agent。

每次真正发起请求前都会重新调用 :func:`app.netguard.validate_node_target`，
缩小 DNS rebinding 的窗口；调用一律携带该节点专属的 node_key。
"""

from __future__ import annotations

import secrets
import time
from typing import Any

import httpx

from .config import Settings
from .models import Node
from .netguard import NodeAddressError, validate_node_target


class NodeClientError(RuntimeError):
    """节点不可达、拒绝了 node_key，或返回了非预期响应。"""


def _base_url(node: Node, settings: Settings) -> str:
    try:
        resolved = validate_node_target(
            node.address,
            node.port,
            node.protocol,
            allow_private=settings.allow_private_nodes,
            allowed_protocols=settings.protocol_list,
            resolve=True,
        )
    except NodeAddressError as exc:
        raise NodeClientError(f"节点地址不再被允许访问: {exc}") from exc
    return resolved.base_url


def _headers(node_key: str) -> dict[str, str]:
    return {"X-Node-Key": node_key}


async def check_health(node: Node, node_key: str, settings: Settings) -> dict[str, Any]:
    base = _base_url(node, settings)
    async with httpx.AsyncClient(timeout=settings.node_health_timeout) as client:
        try:
            resp = await client.get(f"{base}/healthz", headers=_headers(node_key))
        except httpx.HTTPError as exc:
            raise NodeClientError(f"无法连接节点: {exc}") from exc
    if resp.status_code == 401:
        raise NodeClientError("节点拒绝了 node_key")
    if resp.status_code != 200:
        raise NodeClientError(f"节点返回异常状态码 {resp.status_code}")
    return resp.json()


async def measure_ping(
    node: Node, node_key: str, settings: Settings, count: int | None = None
) -> dict[str, Any]:
    base = _base_url(node, settings)
    count = count or settings.ping_count
    latencies: list[float] = []
    async with httpx.AsyncClient(timeout=settings.node_connect_timeout) as client:
        for _ in range(count):
            start = time.perf_counter()
            try:
                resp = await client.get(f"{base}/ping", headers=_headers(node_key))
            except httpx.HTTPError as exc:
                raise NodeClientError(f"ping 失败: {exc}") from exc
            if resp.status_code == 401:
                raise NodeClientError("节点拒绝了 node_key")
            if resp.status_code not in (200, 204):
                raise NodeClientError(f"节点返回异常状态码 {resp.status_code}")
            latencies.append((time.perf_counter() - start) * 1000)
    return {
        "count": len(latencies),
        "min_ms": min(latencies),
        "avg_ms": sum(latencies) / len(latencies),
        "max_ms": max(latencies),
        "jitter_ms": max(latencies) - min(latencies),
    }


async def measure_download(
    node: Node, node_key: str, settings: Settings, num_bytes: int | None = None
) -> dict[str, Any]:
    base = _base_url(node, settings)
    num_bytes = min(num_bytes or settings.default_download_bytes, settings.max_test_bytes)
    total = 0
    start = time.perf_counter()
    async with httpx.AsyncClient(timeout=settings.node_request_timeout) as client:
        try:
            async with client.stream(
                "GET",
                f"{base}/download",
                params={"bytes": num_bytes},
                headers=_headers(node_key),
            ) as resp:
                if resp.status_code == 401:
                    raise NodeClientError("节点拒绝了 node_key")
                if resp.status_code != 200:
                    raise NodeClientError(f"节点返回异常状态码 {resp.status_code}")
                async for chunk in resp.aiter_bytes():
                    total += len(chunk)
        except httpx.HTTPError as exc:
            raise NodeClientError(f"下载测速失败: {exc}") from exc
    elapsed = max(time.perf_counter() - start, 1e-9)
    mbps = (total * 8 / 1_000_000) / elapsed
    return {"bytes": total, "duration_ms": elapsed * 1000, "mbps": mbps}


async def measure_upload(
    node: Node, node_key: str, settings: Settings, num_bytes: int | None = None
) -> dict[str, Any]:
    base = _base_url(node, settings)
    num_bytes = min(num_bytes or settings.default_upload_bytes, settings.max_test_bytes)
    chunk_size = min(settings.stream_chunk_bytes, num_bytes) if num_bytes else 0
    payload_chunk = secrets.token_bytes(chunk_size) if chunk_size else b""

    async def body() -> Any:
        sent = 0
        while sent < num_bytes:
            n = min(len(payload_chunk), num_bytes - sent)
            yield payload_chunk[:n]
            sent += n

    start = time.perf_counter()
    async with httpx.AsyncClient(timeout=settings.node_request_timeout) as client:
        try:
            resp = await client.post(
                f"{base}/upload", headers=_headers(node_key), content=body()
            )
        except httpx.HTTPError as exc:
            raise NodeClientError(f"上传测速失败: {exc}") from exc
    if resp.status_code == 401:
        raise NodeClientError("节点拒绝了 node_key")
    if resp.status_code != 200:
        raise NodeClientError(f"节点返回异常状态码 {resp.status_code}")
    elapsed = max(time.perf_counter() - start, 1e-9)
    mbps = (num_bytes * 8 / 1_000_000) / elapsed
    return {"bytes": num_bytes, "duration_ms": elapsed * 1000, "mbps": mbps}

File: cmd/web/nodeclient.go
package main

import (
	"crypto/rand"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"time"

	"librespeed-service/internal/netguard"
	"librespeed-service/internal/store"
)

// nodeClientError 表示节点不可达、拒绝了 node_key，或返回了非预期响应。
type nodeClientError struct{ msg string }

func (e *nodeClientError) Error() string { return e.msg }

func nodeErr(format string, args ...any) error {
	return &nodeClientError{msg: fmt.Sprintf(format, args...)}
}

func durationFromSeconds(sec float64) time.Duration {
	return time.Duration(sec * float64(time.Second))
}

// nodeBaseURL 每次调用前都重新校验节点地址，缩小 DNS rebinding 的窗口。
func (s *server) nodeBaseURL(node *store.Node) (string, error) {
	resolved, err := netguard.ValidateNodeTarget(node.Address, node.Port, node.Protocol, netguard.ValidateNodeTargetOptions{
		AllowPrivate:     s.settings.AllowPrivateNodes,
		AllowedProtocols: s.settings.AllowedNodeProtocols,
		Resolve:          true,
	})
	if err != nil {
		return "", nodeErr("节点地址不再被允许访问: %s", err.Error())
	}
	return resolved.BaseURL(), nil
}

func (s *server) checkHealth(node *store.Node, nodeKey string) error {
	base, err := s.nodeBaseURL(node)
	if err != nil {
		return err
	}
	client := &http.Client{Timeout: durationFromSeconds(s.settings.NodeHealthTimeout)}
	req, err := http.NewRequest(http.MethodGet, base+"/healthz", nil)
	if err != nil {
		return err
	}
	req.Header.Set("X-Node-Key", nodeKey)
	resp, err := client.Do(req)
	if err != nil {
		return nodeErr("无法连接节点: %s", err.Error())
	}
	defer resp.Body.Close()
	io.Copy(io.Discard, resp.Body)
	if resp.StatusCode == http.StatusUnauthorized {
		return nodeErr("节点拒绝了 node_key")
	}
	if resp.StatusCode != http.StatusOK {
		return nodeErr("节点返回异常状态码 %d", resp.StatusCode)
	}
	return nil
}

func (s *server) measurePing(node *store.Node, nodeKey string, count int) (pingResult, error) {
	base, err := s.nodeBaseURL(node)
	if err != nil {
		return pingResult{}, err
	}
	if count <= 0 {
		count = s.settings.PingCount
	}
	client := &http.Client{Timeout: durationFromSeconds(s.settings.NodeConnectTimeout)}

	latencies := make([]float64, 0, count)
	for i := 0; i < count; i++ {
		start := time.Now()
		req, err := http.NewRequest(http.MethodGet, base+"/ping", nil)
		if err != nil {
			return pingResult{}, err
		}
		req.Header.Set("X-Node-Key", nodeKey)
		resp, err := client.Do(req)
		if err != nil {
			return pingResult{}, nodeErr("ping 失败: %s", err.Error())
		}
		io.Copy(io.Discard, resp.Body)
		resp.Body.Close()
		if resp.StatusCode == http.StatusUnauthorized {
			return pingResult{}, nodeErr("节点拒绝了 node_key")
		}
		if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNoContent {
			return pingResult{}, nodeErr("节点返回异常状态码 %d", resp.StatusCode)
		}
		latencies = append(latencies, time.Since(start).Seconds()*1000)
	}

	minV, maxV, sum := latencies[0], latencies[0], 0.0
	for _, v := range latencies {
		if v < minV {
			minV = v
		}
		if v > maxV {
			maxV = v
		}
		sum += v
	}
	return pingResult{
		Count:    len(latencies),
		MinMs:    minV,
		AvgMs:    sum / float64(len(latencies)),
		MaxMs:    maxV,
		JitterMs: maxV - minV,
	}, nil
}

func (s *server) measureDownload(node *store.Node, nodeKey string, numBytes int64) (transferResult, error) {
	base, err := s.nodeBaseURL(node)
	if err != nil {
		return transferResult{}, err
	}
	if numBytes <= 0 {
		numBytes = s.settings.DefaultDownloadBytes
	}
	if numBytes > s.settings.MaxTestBytes {
		numBytes = s.settings.MaxTestBytes
	}

	client := &http.Client{Timeout: durationFromSeconds(s.settings.NodeRequestTimeout)}
	u := base + "/download?" + url.Values{"bytes": {strconv.FormatInt(numBytes, 10)}}.Encode()
	req, err := http.NewRequest(http.MethodGet, u, nil)
	if err != nil {
		return transferResult{}, err
	}
	req.Header.Set("X-Node-Key", nodeKey)

	start := time.Now()
	resp, err := client.Do(req)
	if err != nil {
		return transferResult{}, nodeErr("下载测速失败: %s", err.Error())
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusUnauthorized {
		return transferResult{}, nodeErr("节点拒绝了 node_key")
	}
	if resp.StatusCode != http.StatusOK {
		return transferResult{}, nodeErr("节点返回异常状态码 %d", resp.StatusCode)
	}
	total, err := io.Copy(io.Discard, resp.Body)
	if err != nil {
		return transferResult{}, nodeErr("下载测速失败: %s", err.Error())
	}
	elapsed := time.Since(start).Seconds()
	if elapsed <= 0 {
		elapsed = 1e-9
	}
	mbps := (float64(total) * 8 / 1_000_000) / elapsed
	return transferResult{Bytes: total, DurationMs: elapsed * 1000, Mbps: mbps}, nil
}

func (s *server) measureUpload(node *store.Node, nodeKey string, numBytes int64) (transferResult, error) {
	base, err := s.nodeBaseURL(node)
	if err != nil {
		return transferResult{}, err
	}
	if numBytes <= 0 {
		numBytes = s.settings.DefaultUploadBytes
	}
	if numBytes > s.settings.MaxTestBytes {
		numBytes = s.settings.MaxTestBytes
	}

	chunkSize := s.settings.StreamChunkBytes
	if int64(chunkSize) > numBytes {
		chunkSize = int(numBytes)
	}
	chunk := make([]byte, chunkSize)
	if chunkSize > 0 {
		if _, err := rand.Read(chunk); err != nil {
			return transferResult{}, err
		}
	}

	client := &http.Client{Timeout: durationFromSeconds(s.settings.NodeRequestTimeout)}
	req, err := http.NewRequest(http.MethodPost, base+"/upload", &repeatReader{chunk: chunk, remaining: numBytes})
	if err != nil {
		return transferResult{}, err
	}
	req.Header.Set("X-Node-Key", nodeKey)
	req.ContentLength = numBytes

	start := time.Now()
	resp, err := client.Do(req)
	if err != nil {
		return transferResult{}, nodeErr("上传测速失败: %s", err.Error())
	}
	defer resp.Body.Close()
	io.Copy(io.Discard, resp.Body)
	if resp.StatusCode == http.StatusUnauthorized {
		return transferResult{}, nodeErr("节点拒绝了 node_key")
	}
	if resp.StatusCode != http.StatusOK {
		return transferResult{}, nodeErr("节点返回异常状态码 %d", resp.StatusCode)
	}
	elapsed := time.Since(start).Seconds()
	if elapsed <= 0 {
		elapsed = 1e-9
	}
	mbps := (float64(numBytes) * 8 / 1_000_000) / elapsed
	return transferResult{Bytes: numBytes, DurationMs: elapsed * 1000, Mbps: mbps}, nil
}

// repeatReader 循环输出 chunk 的内容，总长度截止到 remaining；用于生成上传测速的
// 请求体，避免为大体积测试一次性分配整块内存。
type repeatReader struct {
	chunk     []byte
	remaining int64
	pos       int
}

func (r *repeatReader) Read(p []byte) (int, error) {
	if r.remaining <= 0 || len(r.chunk) == 0 {
		return 0, io.EOF
	}
	max := len(p)
	if int64(max) > r.remaining {
		max = int(r.remaining)
	}
	total := 0
	for total < max {
		n := copy(p[total:max], r.chunk[r.pos:])
		total += n
		r.pos = (r.pos + n) % len(r.chunk)
	}
	r.remaining -= int64(total)
	return total, nil
}
